File: api/embed.js
const { spawn } = require('child_process');
const path = require('path');

module.exports = async (req, res) => {
  try {
    const { embed_url, referrer } = req.query;

    if (!embed_url || !referrer) {
      return res.status(400).json({
        error: 'Missing required parameters: embed_url and referrer are required'
      });
    }

    // Call rabbit.js script for decryption
    const scriptPath = path.join(__dirname, 'rabbit.js');
    const child = spawn('node', [
      scriptPath,
      `--embed-url=${embed_url}`,
      `--referrer=${referrer}`
    ]);

    let output = '';
    let error = '';

    child.stdout.on('data', (data) => {
      output += data;
    });

    child.stderr.on('data', (data) => {
      error += data;
    });

    await new Promise((resolve, reject) => {
      child.on('close', (code) => {
        if (code !== 0) {
          reject(new Error(error || 'Decryption failed'));
        } else {
          resolve();
        }
      });
    });

    // Parse the output
    const result = JSON.parse(output);
    
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Content-Type', 'application/json');
    return res.status(200).json(result);
  } catch (error) {
    console.error('Error:', error);
    return res.status(500).json({
      error: error.message || 'Internal server error'
    });
  }
};

File: api/embed.rs
use vercel_runtime::{run, Body, Error, Request, Response, StatusCode};
use serde::{Deserialize, Serialize};
use std::{process::Command, path::Path};

#[derive(Serialize, Deserialize, Debug)]
pub struct EmbedSources {
    pub sources: Vec<Source>,
    pub tracks: Vec<Track>,
    pub t: u32,
    pub server: u32,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Source {
    pub file: String,
    #[serde(rename = "type")]
    pub source_type: String,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Track {
    pub file: String,
    pub label: String,
    pub kind: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default: Option<bool>,
}

#[tokio::main]
async fn main() -> Result<(), Error> {
    run(handler).await
}

pub async fn handler(req: Request) -> Result<Response<Body>, Error> {
    // Parse query parameters
    let url = req.uri().query()
        .and_then(|q| q.split('&')
            .find(|p| p.starts_with("embed_url="))
            .map(|p| p.split('=').nth(1).unwrap_or("")))
        .ok_or_else(|| Error::from("Missing embed_url parameter"))?;

    let referrer = req.uri().query()
        .and_then(|q| q.split('&')
            .find(|p| p.starts_with("referrer="))
            .map(|p| p.split('=').nth(1).unwrap_or("")))
        .ok_or_else(|| Error::from("Missing referrer parameter"))?;

    // Get the current file's directory
    let current_dir = std::env::current_dir()
        .map_err(|e| Error::from(format!("Failed to get current directory: {}", e)))?;
    
    // Resolve path to rabbit.js
    let script_path = current_dir.join("api").join("rabbit.js");
    
    // Call Node.js script for decryption
    let output = Command::new("node")
        .arg(&script_path)
        .arg(format!("--embed-url={}", url))
        .arg(format!("--referrer={}", referrer))
        .current_dir(script_path.parent().unwrap()) // Set working directory to script's location
        .output()
        .map_err(|e| Error::from(format!("Failed to execute command: {}", e)))?;

    // Check for stderr output
    if !output.stderr.is_empty() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        println!("Script stderr: {}", stderr);
    }

    let parsed_output = String::from_utf8_lossy(&output.stdout).trim().to_string();
    println!("Script output: {}", parsed_output);

    let embed_json: EmbedSources = serde_json::from_str(&parsed_output)
        .map_err(|e| Error::from(format!("Failed to parse JSON: {}", e)))?;

    // Return the response
    Ok(Response::builder()
        .status(StatusCode::OK)
        .header("Content-Type", "application/json")
        .header("Access-Control-Allow-Origin", "*")
        .body(Body::Text(serde_json::to_string(&embed_json)?))?)
}
